File: ntrade/domain/analytics/range_bars.py
"""Range bar generator — price-based bars from 1m OHLCV (Valentini scalper).

A range bar closes when price travels a fixed ``range_size`` (``high - low
>= range_size``), regardless of how many 1m candles that took. This is the
bar primitive Valentini's scalper works on: time-based noise is filtered out
and only price action matters.

Dhan serves only time-based history, so we simulate a canonical path through
each 1m candle (open -> high -> low -> close, or the mirrored order for a
bearish candle) and close a bar the moment its span fills. Candle volume is
distributed proportionally to the path segments each bar consumed — a
documented approximation (no trade tape on Dhan).

Auto range: ``ATR(14)`` scaled to a clean tick grid when ``tick_size`` is
given.
"""

from __future__ import annotations

import pandas as pd

from ntrade.domain.analytics.indicators import atr

_BAR_COLUMNS = ["timestamp", "open", "high", "low", "close", "volume", "is_complete"]


def _empty() -> pd.DataFrame:
    return pd.DataFrame(columns=_BAR_COLUMNS)


def calc_auto_range(df: pd.DataFrame, atr_period: int = 14, multiplier: float = 1.0,
                    tick_size: float | None = None) -> float:
    """ATR-derived range size, optionally rounded to a tick grid.

    Falls back to a sane 10.0 default when ATR cannot be computed (empty
    frame / all-NaN). Never returns <= 0.
    """
    raw = 10.0
    if df is not None and not df.empty:
        a = atr(df, atr_period)
        if len(a) and pd.notna(a.iloc[-1]) and a.iloc[-1] > 0:
            raw = float(a.iloc[-1]) * multiplier
    if tick_size and tick_size > 0:
        raw = max(tick_size, round(raw / tick_size) * tick_size)
    return max(float(raw), 0.5)


def _candle_path(o: float, h: float, lo: float, c: float) -> list[float]:
    """Canonical path through one candle: O -> extreme -> other -> C.

    Bullish: O -> H -> L -> C  (sweep up, wick down, close mid)
    Bearish: O -> L -> H -> C  (sweep down, wick up, close mid)
    """
    return [o, h, lo, c] if c >= o else [o, lo, h, c]


def build_range_bars(df: pd.DataFrame, range_size: float | None = None,
                     atr_period: int = 14, multiplier: float = 1.0,
                     tick_size: float | None = None) -> pd.DataFrame:
    """Convert a 1m OHLCV frame into range bars.

    Returns a frame with columns ``timestamp/open/high/low/close/volume/
    is_complete``. Bars are labelled with the timestamp of the last source
    candle they consumed (right edge). The final open bar is marked
    ``is_complete=False``; earlier bars are complete.
    """
    if df is None or df.empty or "open" not in df:
        return _empty()
    size = range_size if range_size and range_size > 0 else calc_auto_range(
        df, atr_period=atr_period, multiplier=multiplier, tick_size=tick_size)

    bars: list[dict] = []
    cur: dict | None = None  # open/high/low/volume/ts of the in-progress bar

    for _, row in df.iterrows():
        o, h, lo, c = (float(row["open"]), float(row["high"]),
                       float(row["low"]), float(row["close"]))
        raw_vol = row.get("volume", 0)
        vol = float(raw_vol) if not pd.isna(raw_vol) else 0.0
        path = _candle_path(o, h, lo, c)
        segments = len(path) - 1
        vol_per_seg = vol / segments if segments else 0.0
        ts = row["timestamp"]
        for i in range(1, len(path)):
            px = path[i]
            if cur is None:
                # New bar starts at the segment's origin price.
                cur = {"open": path[i - 1], "high": max(path[i - 1], px),
                       "low": min(path[i - 1], px), "close": px,
                       "volume": vol_per_seg, "ts": ts}
            else:
                cur["high"] = max(cur["high"], px)
                cur["low"] = min(cur["low"], px)
                cur["close"] = px
                cur["volume"] += vol_per_seg
                cur["ts"] = ts
            if cur["high"] - cur["low"] >= size:
                bars.append(cur)
                cur = None

    trailing_partial = cur is not None  # last source candle left a bar open
    if trailing_partial:  # trailing incomplete bar (close already set)
        bars.append(cur)

    if not bars:
        return _empty()
    out = pd.DataFrame(bars, columns=["open", "high", "low", "close", "volume", "ts"])
    # Only the trailing partial bar is incomplete — if the final source candle
    # completed a bar exactly (trailing_partial False), every bar is complete.
    is_complete = [True] * len(bars)
    if trailing_partial:
        is_complete[-1] = False
    out["is_complete"] = is_complete
    out["timestamp"] = out["ts"]
    out["volume"] = out["volume"].round(4)
    return out[_BAR_COLUMNS].reset_index(drop=True)


def swing_bias(bars: pd.DataFrame) -> str | None:
    """Directional vote from the last two COMPLETED range bars.

    Bullish when the latest completed bar makes a higher high AND a higher low
    than the prior; bearish on a lower high + lower low; ``None`` otherwise
    (fewer than two completed bars, or an inside/outside bar — no vote).
    """
    if bars is None or bars.empty or "is_complete" not in bars:
        return None
    done = bars[bars["is_complete"].astype(bool)]
    if len(done) < 2:
        return None
    p, c = done.iloc[-2], done.iloc[-1]
    ph, pl = float(p["high"]), float(p["low"])
    ch, cl = float(c["high"]), float(c["low"])
    if ch > ph and cl > pl:
        return "BUY"
    if ch < ph and cl < pl:
        return "SELL"
    return None

File: ui/src/lib/rangeBars.ts
/**
 * Range bars — price-based bars from 1m OHLCV, mirroring
 * `ntrade/domain/analytics/range_bars.py` (the primitive the Valentini
 * scalper actually trades on). A range bar closes when price travels a fixed
 * `rangeSize` (high - low >= rangeSize) regardless of how many 1m candles
 * that took; each candle is traversed along a canonical path
 * (O→H→L→C for bullish, O→L→H→C for bearish) with volume distributed
 * proportionally to the path segments each bar consumed.
 */

import type { Candle } from '../types/market'

/** A range bar — same shape as a candle, plus the completeness flag. */
export interface RangeBar extends Candle {
  isComplete: boolean
}

/**
 * ATR(14) series (Wilder-style ewm, alpha = 1/period). NaN until enough bars
 * exist (min_periods = period), matching the backend's pandas ewm.
 */
export function atrSeries(candles: Candle[], period = 14): number[] {
  // pandas ewm(alpha=1/period, min_periods=period, adjust=True).mean() over
  // the true-range series — the adjust=True (default) form is a normalized
  // weighted mean y_t = Σ_{j≤t} (1-α)^(t-j) x_j / Σ_{j≤t} (1-α)^j, NOT the
  // Wilder recursion (adjust=False). The two diverge ~25% for the first
  // 30-50 bars; matching pandas keeps the ATR floor (and thus SL distance /
  // value-edge / reversal thresholds) in lockstep with the backend.
  const out: number[] = []
  const alpha = 1 / period
  let prevClose = Number.NaN
  let x = 0 // Σ (1-α)^k over all x_j (renormalized on the fly, see below)
  let w = 0 // Σ (1-α)^k weights
  const trs: number[] = []
  for (let i = 0; i < candles.length; i++) {
    const c = candles[i]
    const tr = Number.isNaN(prevClose)
      ? c.high - c.low
      : Math.max(c.high - c.low, Math.abs(c.high - prevClose), Math.abs(c.low - prevClose))
    prevClose = c.close
    trs.push(tr)
  }
  // Adjusted EWMA (pandas default): y_t = (x_t + (1-a)x_{t-1} + ... + (1-a)^t x_0) / (1 + (1-a) + ... + (1-a)^t).
  for (let t = 0; t < trs.length; t++) {
    const a = trs[t]
    w = 1 + (1 - alpha) * w
    x = x * (1 - alpha) + a
    out.push(t >= period - 1 ? x / w : Number.NaN)
  }
  return out
}

/**
 * Convert a user-entered range size (in ticks) to a price range. `null` /
 * missing / non-positive inputs return undefined → the caller falls back to
 * the auto ATR size. The price range is exact (ticks × tick_size), already
 * on the contract's tick grid.
 */
export function rangeSizeFromTicks(ticks: number | null | undefined, tickSize?: number): number | undefined {
  if (ticks && ticks > 0 && tickSize && tickSize > 0) return ticks * tickSize
  return undefined
}

/**
 * ATR-derived range size, optionally rounded to a tick grid. Returns `null`
 * when ATR cannot be computed (too few bars) — the caller must render an
 * "insufficient data" state instead of silently building bars with a magic
 * default size.
 */
export function calcAutoRange(
  candles: Candle[],
  atrPeriod = 14,
  multiplier = 1,
  tickSize?: number,
): number | null {
  if (candles.length === 0) return null
  const a = atrSeries(candles, atrPeriod)
  const last = a[a.length - 1]
  if (!Number.isFinite(last) || !(last > 0)) return null
  let raw = last * multiplier
  if (tickSize && tickSize > 0) raw = Math.max(tickSize, Math.round(raw / tickSize) * tickSize)
  return Math.max(raw, 0.5)
}

/**
 * Convert a 1m OHLCV frame into range bars. Each bar is labelled with the
 * timestamp of the last source candle it consumed (right edge); the trailing
 * partial bar is marked `isComplete: false`. An explicit `rangeSize` beats
 * the auto ATR size (mirrors the backend's `build_range_bars`).
 */
export function buildRangeBars(
  candles: Candle[],
  rangeSize?: number,
  opts: { atrPeriod?: number; tickSize?: number } = {},
): RangeBar[] {
  if (candles.length === 0) return []
  const size = rangeSize && rangeSize > 0
    ? rangeSize
    : calcAutoRange(candles, opts.atrPeriod ?? 14, 1, opts.tickSize)
  // Auto ATR(14) needs enough bars — without a computable size there is no
  // honest range to build; the caller shows an "insufficient data" state.
  if (size == null) return []

  const bars: RangeBar[] = []
  let cur: RangeBar | null = null

  for (const c of candles) {
    const o = c.open
    const h = c.high
    const lo = c.low
    const cl = c.close
    const vol = c.volume || 0
    const path = cl >= o ? [o, h, lo, cl] : [o, lo, h, cl]
    const segments = path.length - 1
    const volPerSeg = segments ? vol / segments : 0
    for (let i = 1; i < path.length; i++) {
      const px = path[i]
      if (cur === null) {
        cur = {
          time: c.time,
          open: path[i - 1],
          high: Math.max(path[i - 1], px),
          low: Math.min(path[i - 1], px),
          close: px,
          volume: volPerSeg,
          isComplete: false,
        }
      } else {
        cur.high = Math.max(cur.high, px)
        cur.low = Math.min(cur.low, px)
        cur.close = px
        cur.volume += volPerSeg
        cur.time = c.time
      }
      if (cur.high - cur.low >= size) {
        cur.isComplete = true // closed bars are complete; only the trailing
        bars.push(cur)        // partial (created with isComplete: false) is not
        cur = null
      }
    }
  }

  if (cur !== null) bars.push(cur) // trailing partial bar (isComplete: false)

  // lightweight-charts requires strictly increasing times, but two range bars
  // can close within the SAME 1m candle and would share its timestamp (and a
  // duplicate-time candlestick series corrupts the chart painter). Nudge a
  // duplicate forward by 1s — invisible at minute granularity, and the next
  // source candle is ≥60s away, so the series stays strictly increasing.
  for (let i = 1; i < bars.length; i++) {
    if (bars[i].time <= bars[i - 1].time) {
      bars[i] = { ...bars[i], time: bars[i - 1].time + 1 }
    }
  }
  return bars
}

/**
 * Directional vote from the last two COMPLETE range bars (mirror of the
 * backend `swing_bias`): higher high + higher low => BUY, lower high +
 * lower low => SELL, else null (fewer than two complete bars, or an
 * inside/outside bar — no vote).
 */
export function swingBias(bars: RangeBar[]): 'BUY' | 'SELL' | null {
  const done = bars.filter((b) => b.isComplete)
  if (done.length < 2) return null
  const p = done[done.length - 2]
  const c = done[done.length - 1]
  if (c.high > p.high && c.low > p.low) return 'BUY'
  if (c.high < p.high && c.low < p.low) return 'SELL'
  return null
}
